### PR TITLE
Expose activation working set as a metric

### PR DESCRIPTION
**Pull request by ahydrax**
This should resolve https://github.com/dotnet/orleans/issues/8279.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8291)

### CODE REVIEW

**Comment by ahydrax** [comment]
reviewers fyi: https://github.com/dotnet/diagnostics/issues/3542
I verified that the bug still present in `dotnet-counters` tool however the fix is merged and we just have to wait for new version.
Anyway, we can test correctness of `UpDownCounter` output by using `prometheus-net`.

**Comment by ReubenBond** [comment]
Thank you for opening this! I believe we should use an observable counter for performance reasons, particularly since we are already tracking the count in an interlocked variable.

**Comment by ahydrax** [comment]
Yeah, I was considering using `ObservableUpDownCounter` but then I had to deal with static/nonstatic context. As metrics live in static context we're unable to use observable counters until we made `ActivationWorkingSet` static somehow.

**Comment by ahydrax** [comment]
For sure I could make that `ActivationWorkingSet` initializes static field with observable counter but for me it will look like a bad design.

**Comment by ahydrax** [comment]
The way I see it is to make `CatalogInstruments` nonstatic then add it to DI and inject `ActivationWorkingSet` into it.
cc @ReubenBond

**Comment by ReubenBond** [comment]
Good point, that is a limitation of how metrics work in .NET today. We don't expect coherent results when multiple silos are hosted within a single process and this problem exists with most of the other counters we publish.

If/when .NET adopts a non-static model for counters, we can adjust our approach to accommodate.

An alternative is to also have a global count property

**Comment by ReubenBond** [comment]
> The way I see it is to make `CatalogInstruments` nonstatic then add it to DI and inject `ActivationWorkingSet` into it.
> cc @ReubenBond 

We should probably do this, too.

**Comment by ahydrax** [comment]
Just a couple of observations made:
- Potentionally big performance issue: we allow to observe ConcurrentDictionary.Count property which could acquire more than one lock at once https://github.com/dotnet/orleans/pull/8291/files#diff-afca937af4caf78f3e9f6df7b74bb6f52eee5e1e4cf9a1764b1f710043b0f0a2L66
- Using approach I suggested will require a bigger refactoring than I expected: metrics are tightly coupled with the domain code in both directions. For example `Catalog` wants to use metric counters but on other side some of counters want to observe `Catalog` properties (see link above).

So I decided to fix major issue and proceed with current approach.

**Review by ReubenBond — APPROVED** [review]

File: src/Orleans.Runtime/Catalog/ActivationDirectory.cs [hunk 1]
```diff
@@ -1,27 +1,46 @@
...
+    private readonly ConcurrentDictionary<GrainId, IGrainContext> _activations = new();
+
+    public ActivationDirectory()
     {
-        private readonly ConcurrentDictionary<GrainId, IGrainContext> _activations = new();
+        CatalogInstruments.RegisterActivationCountObserve(() => Count);
+    }
 
-        public int Count => _activations.Count;
+    public int Count => Volatile.Read(ref _activationsCount);
```
Comment (ReubenBond) [inline src/Orleans.Runtime/Catalog/ActivationDirectory.cs:19]: Volatile is not necessary here, by the way: interlocked operations ensure cache coherence already

Comment (ReubenBond) [inline src/Orleans.Runtime/Catalog/ActivationDirectory.cs:19]:
```suggestion
    public int Count => _activationsCount;
```

File: src/Orleans.Runtime/Catalog/ActivationWorkingSet.cs [hunk 2]
```diff
@@ -35,9 +35,10 @@ private class MemberState
             _logger = logger;
             _scanPeriodTimer = asyncTimerFactory.Create(TimeSpan.FromMilliseconds(100), nameof(ActivationWorkingSet) + "." + nameof(MonitorWorkingSet));
             _observers = observers.ToList();
+            CatalogInstruments.RegisterActivationWorkingSetObserve(() => Count);
         }
 
-        public int Count => _activeCount;
+        public int Count => Volatile.Read(ref _activeCount);
```
Comment (ReubenBond) [inline src/Orleans.Runtime/Catalog/ActivationWorkingSet.cs:41]:
```suggestion
        public int Count => _activeCount;
```

**Pull request merged**